crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -1014,89 +1023,144 @@ struct RangeBlockMode<
     iter: Peekable<std::vec::IntoIter<SealedHeader<<Eth::Provider as HeaderProvider>::Header>>>,
     next: VecDeque<ReceiptBlockResult<Eth::Provider>>,
     max_range: usize,
+    // Stream of ongoing receipt fetching tasks
+    pending_tasks: FuturesOrdered<ReceiptFetchFuture<Eth::Provider>>,
 }
 
 impl<
         Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
     > RangeBlockMode<Eth>
 {
     async fn next(&mut self) -> Result<Option<ReceiptBlockResult<Eth::Provider>>, EthFilterError> {
+        // First, try to return any already processed result
         if let Some(result) = self.next.pop_front() {
             return Ok(Some(result));
         }
 
-        let Some(next_header) = self.iter.next() else {
-            return Ok(None);
-        };
+        // Next, try to get a completed task result if there are pending tasks
+        if !self.pending_tasks.is_empty() {
+            if let Some(task_result) = self.pending_tasks.next().await {
+                let mut chunk_results = task_result?;
+                if !chunk_results.is_empty() {
+                    let first_result = chunk_results.remove(0);
+                    self.next.extend(chunk_results);
+                    return Ok(Some(first_result));
```

Comment:
this is okay, alternatively we could also wrap this section in a loop but this also works